pulp_2to3_migration/app/pre_migration.py

```diff
@@ -320,29 +321,36 @@ def delete_removed_pulp2_content(content_model):
         last_updated = datetime.utcfromtimestamp(last_updated)
 
         # Query all new relations for that content since the last run
+        base_content_relations = Pulp2RepoContent.objects.filter(
+            pulp2_content_type_id=content_type,
+            pulp2_repository__not_in_plan=False,
+            pulp2_created__gte=last_updated,
+        )
+
         content_relations = (
-            Pulp2RepoContent.objects.filter(
-                pulp2_content_type_id=content_type,
-                pulp2_repository__not_in_plan=False,
-                pulp2_created__gte=last_updated,
-            )
-            .select_related("pulp2_repository")
+            base_content_relations.select_related("pulp2_repository")
             .only(
                 "pulp2_repository",
                 "pulp2_created",
             )
             .order_by("pulp2_created")
         )
 
-        mongo_content_qs = content_model.pulp2.objects(
-            id__in=content_relations.values_list("pulp2_unit_id", flat=True)
-        )
-        batched_mongo_content_qs = mongo_content_qs.only(*mongo_fields).batch_size(
-            batch_size
+        pulp2_unit_id_qs = base_content_relations.order_by("pulp2_unit_id").values_list(
+            "pulp2_unit_id", flat=True
         )
-        pulp2_content_by_id = {
-            record.id: record for record in batched_mongo_content_qs.no_cache()
-        }
+
+        pulp2_content_by_id = {}
+        paginator = Paginator(pulp2_unit_id_qs, batch_size)
+        for page in range(1, paginator.num_pages + 1):
+            mongo_content_qs = content_model.pulp2.objects(
+                id__in=paginator.page(page).object_list
+            )
+            batched_mongo_content_qs = mongo_content_qs.only(*mongo_fields).batch_size(
+                batch_size
```

Comment:
So if I understand correctly, this is effectively manually batching the mongodb query?  Do you know why `.batch_size()` didn't have an impact?

Reply:
Hi @dralley . Please refer to https://bugzilla.redhat.com/show_bug.cgi?id=2074099#c6 for detailed explanation.